### PR TITLE
Simplify the API to read package.json files

### PR DESCRIPTION
**Pull request by paulvanbrenk**
Remove unnecessary complex API.

Replaces interface and 4 derived classes with 2 new methods on a factory class. Didn't change any of the logic in the implementation. That would be a next step which would include perf, improvements to project load.

### CODE REVIEW

File: Nodejs/Product/Npm/PackageJsonFactory.cs [hunk 1]
```diff
@@ -1,14 +1,109 @@
...
 namespace Microsoft.NodejsTools.Npm
 {
-    public class PackageJsonFactory
+    public static class PackageJsonFactory
     {
-        public static IPackageJson Create(IPackageJsonSource source)
+        public static IPackageJson Create(string fullPathToFile)
         {
-            return source.Package == null ? null : new PackageJson(source.Package);
+            if (string.IsNullOrEmpty(fullPathToFile) && !fullPathToFile.EndsWith("package.json"))
```
Comment (amcasey) [inline Nodejs/Product/Npm/PackageJsonFactory.cs]: Should this check be case-insensitive?

Comment (amcasey) [reply Nodejs/Product/Npm/PackageJsonFactory.cs]: Also, I would find a null check clearer than `IsNullOrEmpty`.

File: Nodejs/Product/Npm/PackageJsonFactory.cs [hunk 2]
```diff
@@ -1,14 +1,109 @@
...
+                            return Create(reader);
+                        }
+                    }
+                    catch (PackageJsonException pje)
+                    {
+                        WrapExceptionAndRethrow(pje);
+                    }
+                    catch (Exception exc) when (exc is IOException || exc is UnauthorizedAccessException)
+                    {
+                        if (attempts <= 0) { throw; }
```
Comment (amcasey) [inline Nodejs/Product/Npm/PackageJsonFactory.cs]: Would it make sense to check `attempts` as part of the exception filter?  That would have the advantage of preventing the stack from being unwound.

File: Nodejs/Product/Npm/PackageJsonFactory.cs (hunk 1 above)
Comment (paulvanbrenk) [reply Nodejs/Product/Npm/PackageJsonFactory.cs]: Yes, that should be case-insensitive

File: Nodejs/Product/Npm/SPI/RootPackage.cs [hunk 3]
```diff
@@ -25,7 +25,7 @@ internal class RootPackage : IRootPackage
             {
                 if (packageJsonFile.Length < 260)
                 {
-                    this.PackageJson = PackageJsonFactory.Create(new DirectoryPackageJsonSource(fullPathToRootDirectory));
```
Comment (amcasey) [inline Nodejs/Product/Npm/SPI/RootPackage.cs]: Does the code handle `PackageJson` being null?

File: Nodejs/Product/Npm/PackageJsonFactory.cs [hunk 4]
```diff
@@ -1,14 +1,109 @@
...
+
+The following error occurred:
+
+{0}", ae.Message),
+                    ae);
+            }
+
+            return source == null ? null : new PackageJson(source);
+
+            void WrapExceptionAndRethrow(Exception ex)
```
Comment (amcasey) [inline Nodejs/Product/Npm/PackageJsonFactory.cs]: Why pull this out but not the one for ArgumentException?  I'd probably inline it.

File: Nodejs/Product/Npm/PackageJsonFactory.cs [hunk 5]
```diff
@@ -1,14 +1,109 @@
...
+                    retryInterval *= 2; // exponential backoff
+                }
+            }
+
+            return null;
+
+            void WrapExceptionAndRethrow(Exception ex)
+            {
+                throw new PackageJsonException(
+                    string.Format(CultureInfo.CurrentCulture, @"Error reading package.json at '{0}': {1}", fullPathToFile, ex.Message),
```
Comment (amcasey) [inline Nodejs/Product/Npm/PackageJsonFactory.cs]: Probably redundant to say "package.json" when it's going to appear at the end of the path anyway.

File: Nodejs/Product/Npm/PackageJsonFactory.cs [hunk 6]
```diff
@@ -1,14 +1,109 @@
...
+                    ex);
+            }
+        }
+
+        public static IPackageJson Create(TextReader reader)
+        {
+            object source = null;
+            try
+            {
+                var text = reader.ReadToEnd();
```
Comment (amcasey) [inline Nodejs/Product/Npm/PackageJsonFactory.cs]: I'm pretty sure Newtonsoft can deserialize from a reader without the need for this temporary string.

File: Nodejs/Product/Npm/SPI/RootPackage.cs (hunk 3 above)
Comment (paulvanbrenk) [reply Nodejs/Product/Npm/SPI/RootPackage.cs]: yes

File: Nodejs/Product/Npm/PackageJsonFactory.cs (hunk 6 above)
Comment (paulvanbrenk) [reply Nodejs/Product/Npm/PackageJsonFactory.cs]: nope

File: Nodejs/Product/Npm/PackageJsonFactory.cs [hunk 7]
```diff
@@ -1,14 +1,71 @@
...
+                throw new ArgumentException("Expected full path to 'package.json' file.", nameof(fullPathToFile));
+            }
+
+            if (File.Exists(fullPathToFile))
+            {
+                var retryInterval = 500;
+                var attempts = 5;
+
+                // populate _source with retries for recoverable errors.
+                while (--attempts >= 0)
```
Comment (amcasey) [inline Nodejs/Product/Npm/PackageJsonFactory.cs]: Out of curiosity, why isn't this a `for` loop?

File: Nodejs/Product/Npm/PackageJsonFactory.cs [hunk 8]
```diff
@@ -1,14 +1,71 @@
...
+                {
+                    try
+                    {
+                        using (var fin = new FileStream(fullPathToFile, FileMode.Open, FileAccess.Read, FileShare.Read))
+                        using (var reader = new StreamReader(fin))
+                        {
+                            object source = null;
+
+                            var text = reader.ReadToEnd();
```
Comment (amcasey) [inline Nodejs/Product/Npm/PackageJsonFactory.cs]: If we're just going to read the whole file, would it be easier to call `File.ReadAllText`?

Comment (paulvanbrenk) [reply Nodejs/Product/Npm/PackageJsonFactory.cs]: `File.ReadAllText` doesn't seem to set `FileMode`, `FileAccess`, or `FileShare`.

Comment (amcasey) [reply Nodejs/Product/Npm/PackageJsonFactory.cs]: 1) I'm not that concerned, since it lets go when it's done.
2) I'm pretty sure it does.

**Pull request merged**